File: Sources/CursorMagnifier/CaptureRecoveryPolicy.swift
enum CaptureRecoveryAction {
    case startCapture
    case probePermission
    case stop
}

enum CaptureRecoveryPolicy {
    static func action(
        hasReceivedFrameSinceLaunch: Bool,
        preflightGranted: Bool,
        restartAttempts: Int,
        maxRestartAttempts: Int
    ) -> CaptureRecoveryAction {
        if preflightGranted || hasReceivedFrameSinceLaunch {
            return restartAttempts >= maxRestartAttempts ? .stop : .startCapture
        }
        return .probePermission
    }
}

File: quickhighlight-win/QuickHighlight/Capture/CaptureRecoveryPolicy.cs
namespace QuickHighlight.Capture;

internal enum CaptureRecoveryAction
{
    StartCapture,
    ProbeAvailability,
    Stop
}

internal static class CaptureRecoveryPolicy
{
    public static CaptureRecoveryAction Decide(
        bool captureSupported,
        bool hasReceivedFrame,
        int restartAttempts,
        int maxRestartAttempts)
    {
        if (!captureSupported)
        {
            return CaptureRecoveryAction.ProbeAvailability;
        }

        if (!hasReceivedFrame && restartAttempts >= maxRestartAttempts)
        {
            return CaptureRecoveryAction.ProbeAvailability;
        }

        return CaptureRecoveryAction.StartCapture;
    }
}
